Bug 24751: Holdings REST API

taskula opened the pull request:
https://tiketti.koha-suomi.fi/issues/4313
https://tiketti.koha-suomi.fi/issues/4332
https://bugs.koha-community.org/bugzilla3/show_bug.cgi?id=24751

Sponsored-by: The National Library of Finland

tomascohen committed:
Bug 21334: (KS-BACKPORTED) Add bibliographic content type definitions

This patch adds the following content type definitions, to be used by
endpoint authors implementing MARC-related endpoints:

--------------------------------------
| Content-type             | Alias   |
|--------------------------|---------|
| application/marcxml+xml  | marcxml |
| application/marc-in-json | mij     |
| application/marc         | marc    |
--------------------------------------

Code using this aliases will use them like this:
- Suppose you want to output data in MARCXML format:

my $record = cool_method_that_returns_MARC_RECORD();
return $c->render( status => 200, format => 'marcxml', text => $record->as_xml_record );

Note: 'text' by defaults encodes output as UTF-8. If the encoding is not
UTF-8 you should encode the data on your own.

To test:
- Apply this patches
- Run:
  $ kshell
 k$ prove t/db_dependent/Koha/REST/V1.t
=> SUCCESS: Test pass!
- Sign off :-D

Signed-off-by: Michal Denar <black23@gmail.com>

Signed-off-by: Josef Moravec <josef.moravec@gmail.com>

taskula committed:
Bug 24751: Add holdings OpenAPI definitions

To test:
1. prove t/db_dependent/api/v1.t

taskula committed:
Bug 24751: Add a base for holdings REST API

To test:
1. prove t/db_dependent/api/v1/holdings.t

taskula committed:
Bug 24751: (Koha-Suomi) Support search filters in Koha::Biblio::holdings_full

This simplifies fetching holdings records in REST controller

Only applies to KohaSuomi/Koha

taskula committed:
Bug 24751: (Koha-Suomi) holdings_full should respect 1:1 relationship of holdings <=> holdings_full

By forcing marcxml and marcflavour we can only have one holdings_metadata entry

See discussion at
https://bugs.koha-community.org/bugzilla3/show_bug.cgi?id=22700

Only applies to KohaSuomi/Koha

taskula committed:
Bug 24751: Holdings REST API - list()

This commit adds the following REST API endpoint:
    GET /api/v1/holdings

Supports various query parameters.

To test:
1. prove t/db_dependent/api/v1/holdings.t

taskula committed:
Bug 24751: Holdings REST API - get()

This commit adds the following REST API endpoint:
    GET /api/v1/holdings/xxx

To test:
1. prove t/db_dependent/api/v1/holdings.t

taskula committed:
Bug 24751: Holdings REST API - add()

This commit adds the following REST API endpoint:
    POST /api/v1/holdings

To test:
1. prove t/db_dependent/api/v1/holdings.t

taskula committed:
Bug 24751: Holdings REST API - update()

This commit adds the following REST API endpoint:
    PUT /api/v1/holdings

To test:
1. prove t/db_dependent/api/v1/holdings.t

taskula committed:
Bug 24751: Holdings REST API - delete()

This commit adds the following REST API endpoint:
    DELETE /api/v1/holdings

To test:
1. prove t/db_dependent/api/v1/holdings.t

kivilahtio commented on api/v1/swagger/paths/holdings.json:
```diff
@@ -0,0 +1,352 @@
+        {
+          "name": "holding_id",
+          "in": "query",
+          "description": "Case insensitive starts-with search for internal holding identifier",
```
This is a nice idea, but there are unintended consequences, such as trying to get a record with a specific low digits identifier, but getting all its bigger siblings,
eg.
  find 343 => receive [343, 3431, 3432, 3439, 34312145]
This would make it impossible to get only the specific holding_id

kivilahtio commented on api/v1/swagger/paths/holdings.json:
```diff
@@ -0,0 +1,352 @@
+        {
+          "name": "biblionumber",
+          "in": "query",
+          "description": "Case insensitive starts-with search for internal biblio identifier",
```
Same with all the other case-insensitive primary key lookups.

kivilahtio commented on api/v1/swagger/paths/holdings.json:
```diff
@@ -0,0 +1,352 @@
+        {
+          "name": "timestamp",
+          "in": "query",
+          "description": "Case insensitive starts-with search for date and time this recird was last altered",
```
This is a pretty nice idea.
Thanks to ISO 8601 we can easily filter by year-month-day

kivilahtio commented on api/v1/swagger/paths/holdings.json:
```diff
@@ -0,0 +1,352 @@
+        {
+          "name": "ccode",
+          "in": "query",
+          "description": "Case insensitive starts-with search for authorized value for the collection code for this holding",
```
You could add a 'ccode_lookup' -field to the endpoint definition to enable "starts-with"-matching, or some other variety of lookup algorithm, such as "*value*", Levenshtein, etc.
By default one would get the exact matches.

taskula replied:
Removed starts-with matching completely. Default is now an exact match.

kivilahtio commented on Koha/REST/V1/Holding.pm:
```diff
@@ -35,6 +35,30 @@ use Try::Tiny;
+        }
+
+        for my $filter_param ( keys %$args ) {
+            $filter->{$filter_param} = { LIKE => $args->{$filter_param} . "%" };
```
If the user only wants exact match, this is computationally inefficient in the DB.
There are potentially tens of millions of holdings records in the DB.
A fast access should be the default with optional fuzzy matching.

taskula replied:
Starts-with matching is now only for date and datetime fields.

kivilahtio commented on t/db_dependent/api/v1/holdings.t:
```diff
@@ -45,9 +46,48 @@ my $t = Test::Mojo->new('Koha::REST::V1');
+
+    my $biblionumber = create_biblio( 'test' );
+
+    # test with 1 holding
```
Use subtest here to clearly in human-readable form explain the purpose of these tests. Tests should tell a story. A funny story :)
Use Gherkin if possible.
Also when these tests fail, it is easier to find the failing test from the subtest name/output, than calculating the number of failed assertions.

taskula replied:
Now using subtests.

Regarding human-readable form, I'd argue it's better to add consistency. If you want to change a convention, get an approval from the community and then modify existing tests. Inconsistenct tests suck and it's hard for developers to add new code as there is no clear coding convention.

taskula replied:
P.S. when a test fail, it tells you the line number where the failure occured. I find it the easiest way to locate the point of failure in code.

kivilahtio commented on api/v1/swagger/paths/holdings.json:
```diff
@@ -0,0 +1,352 @@
+            "$ref": "../definitions.json#/holdings"
+          }
+        },
+        "404": {
```
400 is missing. Controller throws it.

taskula replied:
Added

kivilahtio commented on api/v1/swagger/paths/holdings.json:
```diff
@@ -0,0 +1,352 @@
+      },
+      "x-koha-authorization": {
+        "permissions": {
+          "editcatalogue": "add_catalogue"
```
You should add a new set of permissions for holdings CRUD

{holdings => qw(create read update delete)}

It would be better to implement those in the staff client as well, but not really imho.

kivilahtio replied:
Adding new permissions is rather easy.

kivilahtio replied:
The edit_catalogue permission in Koha is truly too large.
It might be desirable to have a specific client to just update holdings-records. eg an acquisitions extension.

kohaputti replied:
@kivilahtio I don't know if this is time usefully spent since KS and NatLibFi are moving to KC Koha where the permissions are differently implemented. IMHO, let's just use the add_catalogue. and also because then we should implement separate "add_catalogue" permission for items and not just holdings records, so this is starting to feel like out of scope.

taskula replied:
Personally I'm fine either way. If (the sponsor) NatLibFi thinks it's not worth it, I'll mark this resolved.

taskula replied:
In fact we already have a separate "edit_items" permission (but not add or delete). So, it's not actually a so far fetched idea to add them for holdings.

kivilahtio commented on Koha/REST/V1/Holding.pm:
```diff
@@ -88,6 +88,38 @@ sub get {
+        my $biblio = Koha::Biblios->find( $body->{biblionumber} );
+        unless ( $biblio ) {
+            return $c->render( status  => 400,
+                openapi => { error => "Biblio with given biblionumber not found" } );
```
This might actually be a "404 Not Found" -error

taskula replied:
Good point. Did some reading on it and the common opinion was to use 404. Fixed.

kivilahtio commented on t/db_dependent/api/v1/holdings.t:
```diff
@@ -117,9 +117,75 @@ subtest 'get() tests' => sub {
+        $tx->req->cookies( { name => 'CGISESSID', value => $session_librarian } );
+        $t->request_ok( $tx )->status_is( 400 )
+          ->json_like('/errors/0/path' => qr/biblioitemnumber/);
+        $holding->{biblioitemnumber} = $biblio->biblioitem->biblioitemnumber;
```
I cant believe this is happening ;)

taskula replied:
What do you suggest? $biblio->biblioitemnumber? The schema does not yet support it and implementing support for it is out of scope of this bug.

kivilahtio commented on t/db_dependent/api/v1/holdings.t:
```diff
@@ -117,9 +117,75 @@ subtest 'get() tests' => sub {
+    $tx->req->cookies( { name => 'CGISESSID', value => $session_patron } );
+    $t->request_ok( $tx )->status_is( 403 );
+
+    subtest 'test required parameters' => sub {
```
Good programming rigour.
The 3 separate test cases should be documented.
subtests or test description.

taskula replied:
Done

kivilahtio commented on t/db_dependent/api/v1/holdings.t:
```diff
@@ -45,9 +46,48 @@ my $t = Test::Mojo->new('Koha::REST::V1');
+    my $holding = create_holding( $biblionumber );
+    my $holdings_full = Koha::Biblios->find( $biblionumber )->holdings_full();
+    $tx = $t->ua->build_tx( GET => '/api/v1/holdings' );
+    $tx->req->cookies( { name => 'CGISESSID', value => $session_patron } );
```
There is a gigaton of repetition all over the place with this specific line of code being repeated ad-infinitum in all REST API tests.
```$tx->req->cookies( { name => 'CGISESSID', value => $session_patron } );```
The old REST API test suite puts the authentication token to the Mojo::UserAgent cookie jar, so you can forget about authentication for every request.

taskula replied:
Repetition sucks but implementing another approach is out of scope of this Bug. However, Koha Community has began using Basic Access authenticaiton in API tests, so that specifying the cookie eventually becomes obsolete.

kivilahtio reviewed (changes requested):
Code review.
Good work Lari, here are some comments.

Especially the performance against a DB of tens of millions of holdings-records should be a top priority.

You could add a performance test to the most demanding endpoints.

kohaputti commented on api/v1/swagger/paths/holdings.json:
```diff
@@ -0,0 +1,352 @@
+        }
+      }
+      ],
+      "consumes": ["application/json"],
```
This doesn't appear to be right since the code processed XML

taskula replied:
Good that you point this out. I suggest we follow Koha [Biblios endpoint RFC](https://wiki.koha-community.org/wiki/Biblios_endpoint_RFC). It states

> Several serialization formats and schemas could be supported

https://wiki.koha-community.org/wiki/Biblios_endpoint_RFC#Metadata_format_handling

In order to support this, we need to backport Bug 21334
https://bugs.koha-community.org/bugzilla3/show_bug.cgi?id=21334

taskula replied:
Done. Force pushed plenty of new code.

POST & PUT now consumes
        "application/marcxml+xml",
        "application/marc-in-json",
        "application/marc"

kohaputti commented:
@taskula: just address the issues @kivilahtio and I pointed out, otherwise this looks good to me!

taskula committed:
Bug 24751: (KS-BACKPORTED) Add MARC::Record::MiJ

Remove this commit in version upgrade

from:
https://github.com/Koha-Community/Koha/commit/f67e23c2c2c71b8378ee6612ca28e7d6ecb71f93

taskula committed:
Bug 24751: (KS-BACKPORTED) Add useful subroutines from Bug 20447

tomascohen committed:
Bug 22194: (KS-BACKPORTED) Add Koha::Exceptions::Metadata

https://github.com/Koha-Community/Koha/commits/7d8b96803f664d86762a6afb966051f7d565c40e/Koha/Exceptions/Metadata.pm

taskula committed:
Bug 24751: (Koha-Suomi) Attempt to get biblionumber from MARC::Record in AddHolding()

If biblionumber is not given as a parameter into the sub, we could still get it
from the given MARC record.

Patch only applies to KohaSuomi/Koha

taskula committed:
Bug 24751: Add controller support for multiple content-types

Tests in later commits.

taskula commented:
Backported some useful stuff, added support for multiple content-types in accordance with https://wiki.koha-community.org/wiki/Biblios_endpoint_RFC

Review pls

kohaputti commented:
@kivilahtio can you do the review?

kivilahtio commented on Koha/Exceptions/Metadata.pm:
```diff
@@ -0,0 +1,69 @@
+    }
+);
+
+sub full_message {
```
This is very good!

taskula commented on api/v1/swagger/paths.json:
```diff
@@ -80,6 +80,12 @@
   "/cities/{cityid}": {
     "$ref": "paths/cities.json#/~1cities~1{cityid}"
   },
+  "/holdings": {
```
I just added it before /holds. paths.json was already messed up before this feature. Sorting should be fixed in another Bug. Out of scope.

taskula replied:
I just added /holdings before /holds. paths.json was already messed up before this feature. Sorting should be fixed in another Bug. Out of scope.

taskula replied:
I just added /holdings before /holds. paths.json was already messed up before this feature. Sorting should be fixed in another Bug. Out of scope.

taskula commented on t/db_dependent/api/v1/holdings.t:
```diff
@@ -0,0 +1,665 @@
+
+    $schema->storage->txn_rollback;
+};
+
```
This is already done in Koha Community.

taskula replied:
Yes. It looks like Koha Community has already done it. But it's a long dependency chain to be backported. I guess we can wait for the next Koha version.

taskula committed:
Bug 24751: Add add|edit|delete_catalogue permissions

taskula commented on api/v1/swagger/paths/holdings.json:
```diff
@@ -0,0 +1,439 @@
+        {
+          "name": "timestamp",
+          "in": "query",
+          "description": "Starts-with search for date and time this recird was last altered",
```
Good eyes! Fixed.

taskula commented on t/db_dependent/api/v1/holdings.t:
```diff
@@ -0,0 +1,665 @@
+    C4::Biblio::UpsertMarcSubfield( $holding_marc, '852', 'c', $branchcode2 );
+    my $record = $holding_marc;
+
+    $tx = $t->ua->build_tx( POST
```
Added.

taskula commented on t/db_dependent/api/v1/holdings.t:
```diff
@@ -0,0 +1,665 @@
+
+my $t = Test::Mojo->new('Koha::REST::V1');
+my $tx;
+
```
There is. "test datecreated starts-with search".

taskula commented on api/v1/swagger/definitions/holding.json:
```diff
@@ -48,13 +48,13 @@
     "metadata": {
-      "type": ["string", "null"],
-      "description": "Metadata"
+      "$ref": "holding/metadata.json"
```
>  It seems something similar is already done in definitions.json.
It seems there is a bug with what I was referring to. definitions.json has "availability/biblio" and "availability/item" but it's not valid usage.

I suggest we just stick to what I added in the first place. It works and I don't think it's horribly wrong.

taskula commented:
Fixes from reviews applied. Another review round?

kohaputti commented:
Just checked the code. Looks good to me, I'm going to merge this now. Thanks!
